### PR TITLE
[lstm] implement calcGradient for bidirectional lstm @open sesame 05/06 15:07

### PR DESCRIPTION
**Pull request by lhs8928**
- Implement calcGradient for bidirectional lstm
 - Added test case for bidirectional lstm

Self evaluation:

Build test: [X]Passed [ ]Failed [ ]Skipped
Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: hyeonseok lee <hs89.lee@samsung.com>

### CODE REVIEW

**Review by songgot — APPROVED** [review]
Looks good to me

**Review by jijoongmoon — APPROVED** [review]
LGTM!

File: test/unittest/models/unittest_models_recurrent.cpp [hunk 1]
```diff
@@ -175,46 +175,46 @@ static std::unique_ptr<NeuralNetwork> makeStackedLSTM() {
...
+  for (auto &node : outer_graph) {
+    nn->addLayer(node);
+  }
+
+  nn->setOptimizer(ml::train::createOptimizer("sgd", {"learning_rate = 0.1"}));
+  return nn;
+}
+
+static std::unique_ptr<NeuralNetwork> makeStackedBidirectionalLSTM() {
+  std::unique_ptr<NeuralNetwork> nn(new NeuralNetwork());
```
Comment (jihochu) [inline test/unittest/models/unittest_models_recurrent.cpp:198]: How about using `make_unique()` ?
```suggestion
  auto nn = std::make_unique<NeuralNetwork>();
```

please check the reference: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2013/n3588.txt

> make_unique shares two advantages with make_shared (excluding the third advantage, increased
efficiency). First, unique_ptr<LongTypeName> up(new LongTypeName(args)) must
mention LongTypeName twice, while auto up = make_unique<LongTypeName>(args)
mentions it once. Second, make_unique prevents the unspecified-evaluation-order
leak triggered by expressions like foo(unique_ptr<X>(new X),
unique_ptr<Y>(new Y)). (Following the advice "never say new" is simpler than
"never say new, unless you immediately give it to a named unique_ptr".)

Comment (lhs8928) [reply test/unittest/models/unittest_models_recurrent.cpp]: Thank you for your notice :)

**Review by jihochu — APPROVED** [review]
LGTM

**Comment by jijoongmoon** [comment]
rebase is required to fix the ci.
